service/src/main/java/bio/terra/catalog/service/DatasetService.java

```diff
@@ -12,22 +19,75 @@
 @Service
 public class DatasetService {
   private final DatarepoService datarepoService;
+  private final SamService samService;
+  private final DatasetDao datasetDao;
   private final ObjectMapper objectMapper;
 
   @Autowired
-  public DatasetService(DatarepoService datarepoService, ObjectMapper objectMapper) {
+  public DatasetService(
+      DatarepoService datarepoService,
+      SamService samService,
+      DatasetDao datasetDao,
+      ObjectMapper objectMapper) {
     this.datarepoService = datarepoService;
+    this.samService = samService;
+    this.datasetDao = datasetDao;
     this.objectMapper = objectMapper;
   }
 
   public DatasetsListResponse listDatasets(AuthenticatedUserRequest user) {
     var response = new DatasetsListResponse();
-    for (SnapshotSummaryModel model : datarepoService.getSnapshots(user.getToken())) {
+    for (SnapshotSummaryModel model : datarepoService.getSnapshots(user)) {
       ObjectNode node = objectMapper.createObjectNode();
       node.put("id", model.getId().toString());
       node.put("dct:title", model.getName());
       response.addResultItem(node);
     }
     return response;
   }
+
+  private boolean checkStoragePermission(
+      AuthenticatedUserRequest user, Dataset dataset, SamAction action) {
+    return switch (dataset.storageSystem()) {
+      case TERRA_DATA_REPO -> datarepoService.userHasAction(
+          user, dataset.storageSourceId(), action);
+      case TERRA_WORKSPACE, EXTERNAL -> false;
+    };
+  }
+
+  private void ensureActionPermission(
+      AuthenticatedUserRequest user, Dataset dataset, SamAction action) {
+    if (!checkStoragePermission(user, dataset, action) && !samService.hasAction(user, action)) {
```

Comment:
I think the logic is correct, although maybe it's written in a confusing way. The user has permission if they either have permission through the storage system _or_ if they're an admin user, who can modify any record.

I'll add a comment as the way the admin role is used here isn't obvious. Maybe I'll change the SAM API name from `hasAction()` to `hasGlobalAction()` or something like that since it's being used to check an action for any resource, not a specific resource.

Reply:
Yes, I think a comment or method rename would be useful.